components/search/src/environment_matching.rs

```diff
@@ -81,16 +92,44 @@ fn matches_region_and_locale(
     false
 }
 
-fn matches_distribution(user_distribution_id: &str, environment_distributions: &[String]) -> bool {
-    environment_distributions.is_empty()
-        || environment_distributions.contains(&user_distribution_id.to_string())
+fn matches_version(
+    user_version: &str,
+    environment_min_version: &str,
+    environment_max_version: &str,
+) -> bool {
+    if environment_min_version.is_empty() && environment_max_version.is_empty() {
+        return true;
+    }
+
+    if user_version.is_empty() {
+        return false;
+    }
+
+    if let Ok(user_version) = Version::try_from(user_version) {
+        if !environment_min_version.is_empty() && environment_max_version.is_empty() {
+            if let Ok(min_version) = Version::try_from(environment_min_version) {
+                return user_version >= min_version;
+            }
+        }
+
+        if environment_min_version.is_empty() && !environment_max_version.is_empty() {
+            if let Ok(max_version) = Version::try_from(environment_max_version) {
+                return max_version >= user_version;
+            }
+        }
+
+        if let Ok(min_version) = Version::try_from(environment_min_version) {
+            if let Ok(max_version) = Version::try_from(environment_max_version) {
+                return min_version <= user_version && user_version <= max_version;
+            }
+        }
+    }
+
+    true
 }
 
-fn matches_application(
-    user_application_name: &SearchApplicationName,
-    environment_applications: &[SearchApplicationName],
-) -> bool {
-    environment_applications.is_empty() || environment_applications.contains(user_application_name)
+fn is_empty_or_contains<T: std::cmp::PartialEq>(env_value: &[T], user_value: &T) -> bool {
+    env_value.is_empty() || env_value.contains(user_value)
```

Comment:
Nice use of the `PartialEq` trait bound here!

If you want to get a little fancier—or if you find yourself wanting more generic slice methods like this in the future—you can make an [extension trait](http://xion.io/post/code/rust-extension-traits.html), and have all slices implement it:

```rust
use std::cmp::PartialEq;

trait SliceExt<T: ?Sized> {
    fn is_empty_or_contains(&self, value: &T) -> bool where T: PartialEq;
}

impl<T> SliceExt<T> for [T] {
    fn is_empty_or_contains(&self, value: &T) -> bool where T: PartialEq {
        self.is_empty() || self.contains(value)
    }
}
```

Then, you could use it like this:

```rust
environment.channels.is_empty_or_contains(&user_environment.update_channel)
    && environment.applications.is_empty_or_contains(&user_environment.app_name)
    && environment.device_type.is_empty_or_contains(&user_environment.device_type)
```

And, if you wanted to use your extension methods in multiple modules, you could put the trait inside a separate module—let's say `ext.rs`—and import it into your other modules like any other item:

```rust
use crate::ext::SliceExt;
```